### DISCUSSION TITLE
Provide NX graph (and affected graph) in the devkit

### DISCUSSION THREAD

**Discussion by mpsanchis** [nrwl/nx#23409] [Feature Requests]
- [X] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

I would like to have access to the NX graph and the affected graph in the devkit. If possible, also other utility functions such as "projects that match XYZ"-

## Motivation

I would avoid writing code such as 
```
const affectedProjects = await execa('nx', ['show', 'projects', '--affected', '--json'], {cwd: workspaceRoot})
.then(result => JSON.parse(result.stdout));
```
(unnecessarily calling the NX CLI from a Node project) and instead, import already-existing functions like this:
```
import {affectedProjects} from "@nx/devkit"
...
const affectedProjects = affectedProjects();
```

## Suggested Implementation

I already have access to some functions. In a discord message, I mentioned that I am able to do:
```
import { findMatchingProjects } from 'nx/src/utils/find-matching-projects'
```
however, this is not public API, so I can't rely on a stability from such a function.

You guys would only have to ensure that some of these utility functions that you already use to build the graph, get projects that match patterns, etc, are public API and that you inform of changes when breaking them.

## Alternate Implementations

Duplicating existing functions in the devkit, and making these more stable.

**Comment by FrozenPandaz** [comment]
I don't have any issues with exposing a programmatic API but I'm curious about some use cases people see for this programmatic API. I could imagine some myself but I'd like to hear from others.

Also, what do you think the API would look like? Would it take the same options as the `affected` commands?

**Comment by mpsanchis** [comment]
In this case, I am using the `nx affected` command because I have a concept similar to the release groups: there are X apps/libs that are versioned and released together. What I do is (1) get the Nx affected list, and then (2) expand this group to contain all components that belong to the affected groups. For instance, if only apps A and B are affected, but they belong to different groups G and H, then all the apps/libs from groups G and H (including, but not limited to, A and B) are released.

I did this before having `nx release`. I am currently experimenting with the native `nx release` command and a custom `release-version` generator, to see if I can get the behaviour I want. If I can, I will go with your solution. I don't want to re-invent the wheel if you do this already, but I have had some issues so far (posted in other GitHub issues).

Regarding the API, I guess the same as the CLI would make sense to me 😄

**Reply by raoulkent** [reply]
I think this is very similar to the use-case we have, mentioned here: https://github.com/nrwl/nx/discussions/23438#discussioncomment-11074867

I think the fundamental issue for me is that I have no easy to use API to access the affected graph using only the API's accessible through the devkit.

**Reply by mpsanchis** [reply]
Hi @raoulkent 

I spawn a new process and call nx's CLI (as a lack of better options..)
If you decide to go with the same approach, be aware that the string returned by `npx nx show projects --affected --json` can be either of:
- ["list", "of", "project", "names"], or
- ["some", "files"]\n["list", "of", "project", "names"]

(at least when I tried it)

I had to do a little custom parser to consider this.

Good luck!

**Reply by raoulkent** [reply]
Huge thanks for the heads up and recommendation! I think I'll have to do the same if I am to go ahead with the refactor I'm looking at, since there does not seem to be any activity with regard to this issue as of yet.

But it does not seem to be a unique use-case so hopefully that's something that can end up in the devkit API sooner rather than later.
